### PR TITLE
[Merged by Bors] - Fix hot reloading for read_asset_bytes

### LINKED ISSUE #6780: Hot reloading doesn't trigger on files read via read_asset_bytes
## Bevy version
0.9

## What you did

Tried to write an asset loader that combines multiple files and enable hot reloading:
```rust
use bevy::{
    asset::{AssetLoader, LoadedAsset},
    prelude::*,
    reflect::TypeUuid,
};

fn main() {
    App::new()
        .add_plugins(DefaultPlugins.set(AssetPlugin {
            watch_for_changes: true,
            ..default()
        }))
        .add_asset_loader(TestAssetLoader)
        .add_asset::<TestAsset>()
        .add_startup_system(load)
        .run();
}

fn load(asset_server: Res<AssetServer>) {
    Box::leak(asset_server.load_untyped("test.main").into());
}

#[derive(TypeUuid)]
#[uuid = "2919310c-fbef-4d84-8f68-1bfb8d8b3ca0"]
struct TestAsset;

struct TestAssetLoader;

impl AssetLoader for TestAssetLoader {
    fn load<'a>(
        &'a self,
        _: &'a [u8],
        load_context: &'a mut bevy::asset::LoadContext,
    ) -> bevy::utils::BoxedFuture<'a, Result<(), bevy::asset::Error>> {
        info!("Loading");
        load_context.set_default_asset(LoadedAsset::new(TestAsset));
        Box::pin(async move {
            load_context.read_asset_bytes("secondary").await?;
            Ok(())
        })
    }

    fn extensions(&self) -> &[&str] {
        &["main"]
    }
}
```

## What went wrong

If `test.main` gets touched, the asset gets reloaded, but if `secondary` gets touched, it doesn't.

### PR DESCRIPTION
**Pull request by SpecificProtagonist**
# Objective

Fixes #6780

## Solution

- record the asset path of each watched file
- call `AssetIo::watch_for_changes` in `LoadContext::read_asset_bytes`

---

### Fixed
- fixed hot reloading for `LoadContext::read_asset_bytes`

### Changed
- `AssetIo::watch_path_for_changes` allows watched path and path to reload to differ

## Migration Guide
- for custom `AssetIo`s, differentiate paths to watch and asset paths to reload as a consequence

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
bors try

File: crates/bevy_asset/src/io/mod.rs [hunk 1]
```diff
@@ -66,7 +66,15 @@ pub trait AssetIo: Downcast + Send + Sync + 'static {
     fn get_metadata(&self, path: &Path) -> Result<Metadata, AssetIoError>;
 
     /// Tells the asset I/O to watch for changes recursively at the provided path.
-    fn watch_path_for_changes(&self, path: &Path) -> Result<(), AssetIoError>;
+    /// No-op if `watch_for_changes` hasn't been called yet.
```
Comment (alice-i-cecile) [inline crates/bevy_asset/src/io/mod.rs:69]:
```suggestion
    ///
    /// No-op if `watch_for_changes` hasn't been called yet.
```

**Comment by alice-i-cecile** [comment]
Awesome thank you

File: crates/bevy_asset/src/io/mod.rs [hunk 2]
```diff
@@ -66,7 +66,16 @@ pub trait AssetIo: Downcast + Send + Sync + 'static {
     fn get_metadata(&self, path: &Path) -> Result<Metadata, AssetIoError>;
 
     /// Tells the asset I/O to watch for changes recursively at the provided path.
-    fn watch_path_for_changes(&self, path: &Path) -> Result<(), AssetIoError>;
+    ///
+    /// No-op if `watch_for_changes` hasn't been called yet.
+    /// Otherwise, if `to_watch` gets modified, trigger a reload for `to_reload`.
+    /// Note that multiple `to_reload` paths can map to the same `to_watch` and
+    /// multiple `to_watch` paths can map to the same `to_reload`.
```
Comment (mockersf) [inline crates/bevy_asset/src/io/mod.rs:73]: Could you add more details on why someone calling this method would want to provide two different paths?

File: crates/bevy_asset/src/io/mod.rs [hunk 3]
```diff
@@ -66,7 +66,16 @@ pub trait AssetIo: Downcast + Send + Sync + 'static {
...
+    fn watch_path_for_changes(
+        &self,
+        to_watch: &Path,
+        to_reload: PathBuf,
```
Comment (mockersf) [inline crates/bevy_asset/src/io/mod.rs:77]: It seems they will often be the same path, could to_reload be an option in that case?

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
bors try

**Review by cart — APPROVED** [review]

**Pull request closed without merging**